feat: add `relative-font-units` rule (#133)
eslint/css: docs/rules/relative-font-units.md

# relative-font-units

Enforce the use of relative units for font size.

## Background

The `font-size` property in CSS defines the size of the text. It can be set using:

> 1. Keywords (e.g., small, medium, large).
> 1. Length units (e.g., px, em, rem, pt).
> 1. Percentages (%, relative to the parent element's font size).

Reviewer: ```suggestion
1. Keywords (e.g., `small`, `medium`, `large`).
1. Length units (e.g., `px`, `em`, `rem`, `pt`).
1. Percentages (`%`, relative to the parent element's font size).
```

Author: Done!

> Generally, relative unit such as `rem` or `em` are preferred over absolute ones like `px`, `pt` because of following reasons:

Reviewer 2: ```suggestion
Generally, relative units such as `rem` or `em` are preferred over absolute ones like `px`, `pt` because of the following reasons:
```

- **Responsive Design** - Relative units adapt better to various screen widths and pixel densities (e.g., mobile vs. desktop).
- **Accessibility** - Relative units allow text to scale when users adjust browser settings or zoom levels.
- **Consistency and Scalability** - Using relative units allow for consistent scaling across a whole site.
- **Maintainability and Reusability** - Relative units allow components or utility classes to work well in different contexts.

## Rule Details

This rule enforces the use of relative units for font size.

> ### Options

Reviewer 2: I think it should be h2.

```suggestion
## Options
```

> This rule accepts an option which is an object with following property:

Reviewer 2: ```suggestion
This rule accepts an option which is an object with the following property:
```

> - `allow` (default: `["rem"]`) - Specify an array of relative units that are allowed to be used. You can use following units:

Reviewer 2: ```suggestion
- `allow` (default: `["rem"]`) - Specify an array of relative units that are allowed to be used. You can use the following units:
```

    - **%**: Represents the "percentage" of the parent element’s font size, allowing the text to scale relative to its container.
    - **cap**: Represents the "cap height" (nominal height of capital letters) of the element's font.
    - **ch**: Represents the width or advance measure of the "0" glyph in the element's font.
    - **em**: Represents the calculated font-size of the element.
    - **ex**: Represents the x-height of the element's font.
    - **ic**: Equal to the advance measure of the "水" glyph (CJK water ideograph) in the font.
    - **lh**: Equal to the computed line-height of the element.
    - **rcap**: Equal to the "cap height" of the root element's font.
    - **rch**: Equal to the width or advance measure of the "0" glyph in the root element's font.
    - **rem**: Represents the font-size of the root element.
    - **rex**: Represents the x-height of the root element's font.
    - **ric**: Equal to the ic unit on the root element's font.
    - **rlh**: Equal to the lh unit on the root element's font.

Example of **incorrect** code for default `{ allow: ["rem"] }` option:

> ```css

Reviewer 2: ```suggestion
```css
/* eslint css/relative-font-units: ["error",  { allow: ["rem"] }] */

```

a {
	font-size: 10px;
}

b {
	font-size: 2em;
}

c {
	font-size: small;
}
```

Example of **correct** code for default `{ allow: ["rem"] }` option:

> ```css
> a {

Reviewer 2: ```suggestion
```css
/* eslint css/relative-font-units: ["error",  { allow: ["rem"] }] */

a {
```

	font-size: 2rem;
}

b {
	font-size: 1rem;
	width: 20px;
}

c {
	font-size: var(--foo);
}
```

Font size can also be specified in `font` property:

Example of **correct** code for `{ allow: ["em", "%"] }` option:

> ```css
> a {

Reviewer 2: ```suggestion
```css
/* eslint css/relative-font-units: ["error",  { allow: ["em", "%"] }] */
a {
```

    - **ex**: Represents the x-height of the element's font.
    - **ic**: Equal to the advance measure of the "水" glyph (CJK water ideograph) in the font.
    - **lh**: Equal to the computed line-height of the element.
    - **rcap**: Equal to the "cap height" of the root element's font.
    - **rch**: Equal to the width or advance measure of the "0" glyph in the root element's font.
    - **rem**: Represents the font-size of the root element.
    - **rex**: Represents the x-height of the root element's font.
    - **ric**: Equal to the ic unit on the root element's font.
    - **rlh**: Equal to the lh unit on the root element's font.

Example of **incorrect** code for default `{ allow: ["rem"] }` option:

```css
a {
	font-size: 10px;
}

b {
	font-size: 2em;
}

c {
	font-size: small;
}
```

Example of **correct** code for default `{ allow: ["rem"] }` option:

```css
a {
	font-size: 2rem;
}

b {
	font-size: 1rem;
	width: 20px;
}
```

Font size can also be specified in `font` property:

Example of **correct** code for `{ allow: ["em", "%"] }` option:

```css
a {
	font-size: 2em;
}

b {
	font:
		20% Arial,
		sans-serif;
}
```

> ## When Not to Use It

Author: Done!

		20% Arial,
		sans-serif;
}

c {
	font: Arial var(--foo);
}
```

## When Not to Use It

If your project does not prioritize the use of relative font-size units—such as in cases requiring absolute sizing for print styles, pixel-perfect UI components, or embedded widgets—you may safely disable this rule without impacting your intended design precision.

> ## Prior Art

Reviewer: ```suggestion
## Further Reading
```
